File: modules/terminal-view/android/src/main/java/expo/modules/terminalview/LinkDetector.kt
package expo.modules.terminalview

/**
 * Detects clickable links in terminal output text:
 *   - URLs (http/https)
 *   - File paths (absolute, relative, home-relative)
 *   - Error references (file:line:col format from compilers/linters)
 */
object LinkDetector {

    enum class LinkType {
        URL,
        FILE_PATH,
        ERROR_REF
    }

    data class DetectedLink(
        val text: String,
        val type: LinkType,
        val startIndex: Int,
        val endIndex: Int,
        val filePath: String? = null,
        val line: Int? = null,
        val column: Int? = null
    )

    // URL pattern: http(s)://... until whitespace or common terminal chars.
    // 2026-08-07 on-device QA finding (docs/superpowers/DEFERRED.md): no
    // IGNORE_CASE meant an upper/mixed-case scheme (e.g. `echo
    // "HTTPS://EXAMPLE.COM/PATH"`, common for shell-scripted output) was
    // never detected as a link at all, not merely unclickable.
    private val URL_PATTERN = Regex(
        """https?://[^\s<>"'\])|}\x1b]+""",
        RegexOption.IGNORE_CASE
    )

    // Error reference: file.ext:42 or file.ext:42:10
    // Must have an extension to avoid false positives
    private val ERROR_REF_PATTERN = Regex(
        """(?:^|[\s(])([./~]?[\w./_-]+\.\w+):(\d+)(?::(\d+))?"""
    )

    // File path: starts with /, ./, ../, or ~/
    // Must contain at least one / to distinguish from regular words
    private val FILEPATH_PATTERN = Regex(
        """(?:^|[\s"'(])([/~][\w./_-]*(?:/[\w./_-]+)+)"""
    )

    /**
     * Detect all links in the given text.
     * Returns a list of DetectedLink sorted by start position.
     */
    fun detect(text: String): List<DetectedLink> {
        val links = mutableListOf<DetectedLink>()
        val usedRanges = mutableListOf<IntRange>()

        // 1. URLs first (highest priority)
        for (match in URL_PATTERN.findAll(text)) {
            val link = DetectedLink(
                text = match.value,
                type = LinkType.URL,
                startIndex = match.range.first,
                endIndex = match.range.last + 1
            )
            links.add(link)
            usedRanges.add(match.range)
        }

        // 2. Error references (file:line:col)
        for (match in ERROR_REF_PATTERN.findAll(text)) {
            val range = IntRange(match.groups[1]!!.range.first, match.range.last)
            if (usedRanges.any { it.overlaps(range) }) continue

            val filePath = match.groupValues[1]
            val line = match.groupValues[2].toIntOrNull()
            val column = match.groupValues[3].toIntOrNull()

            links.add(DetectedLink(
                text = text.substring(range),
                type = LinkType.ERROR_REF,
                startIndex = range.first,
                endIndex = range.last + 1,
                filePath = filePath,
                line = line,
                column = column
            ))
            usedRanges.add(range)
        }

        // 3. File paths
        for (match in FILEPATH_PATTERN.findAll(text)) {
            val pathGroup = match.groups[1] ?: continue
            val range = pathGroup.range
            if (usedRanges.any { it.overlaps(range) }) continue

            links.add(DetectedLink(
                text = pathGroup.value,
                type = LinkType.FILE_PATH,
                startIndex = range.first,
                endIndex = range.last + 1,
                filePath = pathGroup.value
            ))
            usedRanges.add(range)
        }

        return links.sortedBy { it.startIndex }
    }

    private fun IntRange.overlaps(other: IntRange): Boolean {
        return this.first <= other.last && other.first <= this.last
    }
}

File: lib/link-detector.ts
/**
 * Detects clickable links (URLs and file paths) in terminal output text.
 */

export type DetectedLink = {
  text: string;
  type: 'url' | 'filepath';
  start: number;
  end: number;
};

// URL pattern: http(s)://... or www.... Case-insensitive: a URL echoed back
// in upper/mixed case (common for shell-scripted output, e.g. `echo
// "HTTPS://EXAMPLE.COM/PATH"`) is just as clickable as a lowercase one — the
// scheme/host are case-insensitive per RFC 3986, and this bare pattern-match
// step doesn't care about the path's case either.
const URL_REGEX = /https?:\/\/[^\s<>"')\]},;]+|www\.[^\s<>"')\]},;]+/gi;

// File path pattern: /path/to/file or ./relative/path (with extension)
const PATH_REGEX = /(?:\/[\w.-]+){2,}(?:\.\w{1,10})?|\.\/[\w./-]+/g;

/**
 * Find all URLs and file paths in a text line.
 */
export function detectLinks(text: string): DetectedLink[] {
  const links: DetectedLink[] = [];
  const seen = new Set<string>();

  // URLs
  let match: RegExpExecArray | null;
  URL_REGEX.lastIndex = 0;
  while ((match = URL_REGEX.exec(text)) !== null) {
    const key = `${match.index}-${match[0]}`;
    if (!seen.has(key)) {
      seen.add(key);
      links.push({
        text: match[0],
        type: 'url',
        start: match.index,
        end: match.index + match[0].length,
      });
    }
  }

  // File paths (only if no URL overlap)
  PATH_REGEX.lastIndex = 0;
  while ((match = PATH_REGEX.exec(text)) !== null) {
    const start = match.index;
    const end = start + match[0].length;
    const overlaps = links.some((l) => start >= l.start && start < l.end);
    if (!overlaps) {
      const key = `${start}-${match[0]}`;
      if (!seen.has(key)) {
        seen.add(key);
        links.push({
          text: match[0],
          type: 'filepath',
          start,
          end,
        });
      }
    }
  }

  return links.sort((a, b) => a.start - b.start);
}

/**
 * Split text into segments: plain text and linked text.
 */
export type TextSegment = {
  text: string;
  link?: DetectedLink;
};

export function segmentText(text: string): TextSegment[] {
  const links = detectLinks(text);
  if (links.length === 0) return [{ text }];

  const segments: TextSegment[] = [];
  let pos = 0;

  for (const link of links) {
    if (link.start > pos) {
      segments.push({ text: text.slice(pos, link.start) });
    }
    segments.push({ text: link.text, link });
    pos = link.end;
  }

  if (pos < text.length) {
    segments.push({ text: text.slice(pos) });
  }

  return segments;
}
